### PR TITLE
TD-1467 Subscribe to notification when delegate is registered or prom…

### PR DESCRIPTION
**Pull request by manishagarwalnhs**
…oted to admin

### JIRA link
https://hee-tis.atlassian.net/browse/TD-1467

### Description
Points addressed in this Commit :
1) Subscribe to notifications when delegate is registered internally or externally.
2) Subscribe to notifications when delegate is promoted to admin

### Screenshots
This is backend change so no screenshot available.

-----
### Developer checks
(Leave tasks unticked if they haven't been appropriate for your ticket.)

I have:
- [x] Run the formatter and made sure there are no IDE errors (see [info on Text Editor settings](https://hee-tis.atlassian.net/wiki/spaces/TP/pages/3546185813/DLS+Dev+Process) to avoid whitespace changes)
- [ ] Written tests for the changes (accessibility tests, unit tests for controller, data services, services, view models, etc)
- [x] Manually tested my work with and without JavaScript
- [ ] Tested any Views or partials created or changed with [Wave Chrome plugin](https://chrome.google.com/webstore/detail/wave-evaluation-tool/jbbplnpkjmmeebjpijfedlgcdilocofh/related) and addressed any valid accessibility issues
- [ ] Updated/added documentation in [Confluence](https://hee-tis.atlassian.net/wiki/spaces/TP/pages/3546939432/DLS+Code) and/or [GitHub Readme](https://github.com/TechnologyEnhancedLearning/DLSV2/blob/master/README.md). List of documentation links added/changed:
  - [doc_1_here](link_1_here)
- [x] Updated my Jira ticket with information about other parts of the system that were touched as part of the MR and have to be sanity tested to ensure nothing’s broken
- [x] Scanned over my pull request in GitHub and addressed any warnings from the GitHub Build and Test checks.

### CODE REVIEW

File: DigitalLearningSolutions.Data/DataServices/NotificationDataService.cs [hunk 1]
```diff
@@ -105,5 +107,24 @@ public IEnumerable<int> GetRoleBasedNotifications(int isCentreManager, int isCon
...
+                connection.Execute(
+                @"INSERT INTO NotificationUsers (
+                    NotificationID,CandidateID)
+                    (SELECT NotificationID,@candidateId FROM NotificationRoles WHERE RoleID = 5)",new {candidateId});
+            }
+
+            if (adminUserId != null)
+            {
+                connection.Execute(
+                @"INSERT INTO NotificationUsers (
```
Comment (kevwhitt-hee) [inline DigitalLearningSolutions.Data/DataServices/NotificationDataService.cs:123]: The logic removed at line 237 of RegistrationDataService.cs is better than this, we should revert this code to continue to use that logic. I had assumed that the subscription to notifications had been missed entirely but I see now that it was only present for admin registrations.

File: DigitalLearningSolutions.Data/DataServices/NotificationDataService.cs [hunk 2]
```diff
@@ -105,5 +107,24 @@ public IEnumerable<int> GetRoleBasedNotifications(int isCentreManager, int isCon
                 ).AsEnumerable();
         }
 
+        public void SubscribeDefaultNotifications(int? candidateId, int? adminUserId)
+        {
+            if (candidateId != null)
+            {
+                connection.Execute(
+                @"INSERT INTO NotificationUsers (
+                    NotificationID,CandidateID)
```
Comment (kevwhitt-hee) [inline DigitalLearningSolutions.Data/DataServices/NotificationDataService.cs:116]: We should join to the Notifications table here and honour the AutoOptIn flag (only insert a record if it is True) as per the existing logic for admin notification subscriptions.

**Review by kevwhitt-hee — CHANGES_REQUESTED** [review]
The previous logic for admin notifications is better than the new logic - we should keep it.
We should also honour the AutoOptIn flag for delegate notifications - see code comments.

File: DigitalLearningSolutions.Data/DataServices/NotificationDataService.cs (hunk 1 above)
Comment (manishagarwalnhs) [reply DigitalLearningSolutions.Data/DataServices/NotificationDataService.cs]: This is addressed.

File: DigitalLearningSolutions.Data/DataServices/NotificationDataService.cs (hunk 2 above)
Comment (manishagarwalnhs) [reply DigitalLearningSolutions.Data/DataServices/NotificationDataService.cs:114]: This is addressed.

**Review by kevwhitt-hee — APPROVED** [review]

**Pull request merged**